tests/unit/test_test_procedures.py

```diff
@@ -115,3 +119,41 @@ def gather_connected_steps(links: dict[str, list[str]], node: str, visited_nodes
     assert (
         step_names == visited_nodes
     ), "Missing entries here indicate a step (or steps) that can't be reached from the root node"
+
+
+@pytest.mark.parametrize("tp_id, tp", ALL_TEST_PROCEDURES)
+def test_procedures_have_required_preconditions(tp_id: str, tp: TestProcedure):
+
+    # Check 'end-device-contents' present
+    enddevice_not_required = [
+        "ALL-01",  # Out-of-band device registration
+        "ALL-02",  # In-band registration during test procedure
+        "OPT-1-IN-BAND",  # In-band device registration during test procedure
+        "OPT-1-OUT-OF-BAND",  # Out-of-band device registration
+        "ALL-04",  # In-band device registration during test procedure
+    ]
+    if tp_id not in enddevice_not_required:
+        assert tp.preconditions is not None
+        assert tp.preconditions.checks is not None
+        assert any([check.type == "end-device-contents" for check in tp.preconditions.checks])
+
+    # Check 'der-settings-contents' present if precondition action involves "opModImpLimW" or "opModGenLimW"
+    # expression parameters.
+    if tp.preconditions is not None and tp.preconditions.actions is not None:
+        for action in tp.preconditions.actions:
+            if action.type == "create-der-control" and action.parameters is not None:
```

Comment:
Is the `action.type == "create-der-control"` still necessary? Shouldn't it just be `if action.parameters is not None`? From there - can't you just do a loop through all parameter keys?